### PR TITLE
LPS-159013 ES module custom element portlets render incorrect <script> node when dragged in the UI

### PR DESCRIPTION
**Pull request by izaera**

### CODE REVIEW

**Comment by izaera** [comment]
ci:test:relevant

**Comment by izaera** [comment]
ci:test:sf

**Comment by izaera** [comment]
ci:test:relevant

File: portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+			(ThemeDisplay)httpServletRequest.getAttribute(
+				WebKeys.THEME_DISPLAY);
+
+		if (urlType == URLType.CSS) {
+			fastLoad = themeDisplay.isThemeCssFastLoad();
+		}
+		else if (urlType == URLType.JAVASCRIPT) {
+			fastLoad = themeDisplay.isThemeJsFastLoad();
+
+			predicate = resource -> !themeDisplay.isIncludedJs(resource);
```
Comment (bryceosterhaus) [inline portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java:411]: I've seen this `predicate` before, but I really have no idea what is does. Could you explain it very briefly or link me to something about the purpose of `predicate`?

File: portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+			visitedURLs = new LinkedHashSet<>();
+
+			httpServletRequest.setAttribute(
+				WebKeys.PORTLET_RESOURCE_STATIC_URLS, visitedURLs);
+		}
+
+		if (fastLoad) {
+			Theme theme = themeDisplay.getTheme();
+
+			return _getComboServletURLs(
```
Comment (bryceosterhaus) [inline portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java:431]: Are early returns okay to use in portal? For some reason I remember bchan or someone not liking them... I could be widely off here though

File: portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+
+		return _getStaticURLs(
+			httpServletRequest, portletResourceAccessors, portlets,
+			visitedURLs);
+	}
+
+	private static void _writeCSSPath(PrintWriter printWriter, String cssPath) {
+		printWriter.print("<link href=\"");
+		printWriter.print(HtmlUtil.escape(cssPath));
+		printWriter.println("\" rel=\"stylesheet\" type=\"text/css\" />");
```
Comment (bryceosterhaus) [inline portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java:450]: Don't we need `data-senna-track="temporary"` here?

Comment (bryceosterhaus) [reply portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java:450]: It was previously there, https://github.com/liferay-frontend/liferay-portal/pull/2580/files#diff-1afd0967f8255757d64743e17acb070730b3094f09a0aa78a9608b4e2b379595L31

**Review by bryceosterhaus — COMMENTED** [review]
This looks pretty great! Way better than our previous implementation

File: portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java (hunk 2 above)
Comment (izaera) [reply portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java:431]: It's now the blessed pattern. "We" even prefer using "if + return then more code" than "if + return else more code"

File: portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java (hunk 3 above)
Comment (izaera) [reply portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java:450]: argh!!! :scream: And there's an ID too. I need to fix it :-(

Comment (izaera) [reply portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java]: Fixed in a new commit

**Comment by izaera** [comment]
ci:test:sf

**Comment by izaera** [comment]
ci:test:relevant

**Comment by izaera** [comment]
ci:forward

**Comment by izaera** [comment]
ci:forward

**Comment by izaera** [comment]
ci:test:relevant

**Comment by izaera** [comment]
ci:forward

**Comment by magjed4289** [comment]
ci:stop

**Comment by magjed4289** [comment]
ci:test:stop

**Comment by magjed4289** [comment]
Just started reviewing :)

**Comment by magjed4289** [comment]
ci:forward

**Comment by izaera** [comment]
ci:forward

**Comment by magjed4289** [comment]
Just started reviewing :)

**Comment by izaera** [comment]
After review with QA (@magjed4289 ) this PR is safe to be manually forwarded.

**Comment by magjed4289** [comment]
Test failures are unrelated. OK to manually forward ✅

**Comment by izaera** [comment]
Sent as https://github.com/brianchandotcom/liferay-portal/pull/121205

**Pull request closed without merging**